src/main/scala/cognite/spark/v1/MetricsSource.scala

```diff
@@ -2,22 +2,36 @@ package org.apache.spark.datasource
 
 import cats.Eval
 import com.codahale.metrics._
-import org.apache.spark._
-
 import java.util.concurrent.ConcurrentHashMap
+import org.apache.spark._
 
 class MetricsSource {
   // Add metricNamespace to differentiate with spark system metrics.
   // Keeps track of all the Metric instances that are being published
   val metricsMap = new ConcurrentHashMap[String, Eval[Counter]]
 
-  def getOrCreateCounter(metricNamespace: String, metricName: String): Counter = {
+  def getOrCreateCounter(metricNamespace: String, name: String): Counter = {
+    val ctx = Option(TaskContext.get())
```

Comment:
is that TaskContext coming from a thread-local horror land? :thinking:

Reply:
if it is then val->def change makes sense but would be better than to move that stuff around, let MetricSource.getCounter take TaskContext as parameter and pass it in CdfRelation and even there maybe also take it as a parameter and require filling it at metric update sites like
```scala
        if (config.collectMetrics) {
          itemsRead.inc()
        }
```
so we definitely have more control on which thread it gets executed on (datasource repo uses cats-effect in various places, to switches between threads are quite likely)

Reply:
Got it. Yeah, I suppose that can be done. Then the counters can become fields again (referring to the comments above 👆 )